src/flag_gems/ops/conv2d.py

```diff
@@ -0,0 +1,311 @@
+    input_pointer += (
+        input_n_stride * in_n_point_value + input_c_stride * pid_group * kernel_c
+    )[:, None]
+    weight_pointer += (
+        weight_n_stride * output_c_offset
+        + weight_n_stride * pid_group * out_per_group_c
+    )[None, :]
+
+    accum = tl.zeros((BLOCK_NI_HO_WO, BLOCK_CO), dtype=tl.float32)
+
+    for h in range(kernel_height):
+        for w in range(kernel_width):
+            for c in range(0, kernel_c, BLOCK_CI):
+                input_c_offset = c + tl.arange(0, BLOCK_CI)
+                input_height_offset = (
+                    h - padding_height + stride_height * output_height_point_value
+                )
+                input_width_offset = (
+                    w - padding_width + stride_width * output_width_point_value
+                )
+
+                curr_input_pointer = (
+                    input_pointer
+                    + (input_c_stride * input_c_offset)[None, :]
+                    + (input_height_stride * input_height_offset)[:, None]
+                    + (input_width_stride * input_width_offset)[:, None]
+                )
+                curr_weight_pointer = (
+                    weight_pointer
+                    + (weight_c_stride * input_c_offset)[:, None]
+                    + (weight_height_stride * h)[None, :]
+                    + (weight_width_stride * w)[None, :]
+                )
+
+                input_mask = (
+                    (in_n_point_value < in_n)[:, None]
+                    & (input_c_offset < kernel_c)[None, :]
+                    & (0 <= input_height_offset)[:, None]
+                    & (input_height_offset < in_height)[:, None]
+                    & (0 <= input_width_offset)[:, None]
+                    & (input_width_offset < in_width)[:, None]
+                )
+                weight_mask = (input_c_offset < kernel_c)[:, None] & (
+                    output_c_offset < out_per_group_c
+                )[None, :]
+
+                input_block = tl.load(curr_input_pointer, mask=input_mask)
+                weight_block = tl.load(curr_weight_pointer, mask=weight_mask)
+
+                accum += tl.dot(input_block, weight_block, allow_tf32=False)
+
+    output_pointer += (
+        (output_n_stride * in_n_point_value)[:, None]
+        + (output_c_stride * (pid_group * out_per_group_c + output_c_offset))[None, :]
+        + (output_height_stride * output_height_point_value)[:, None]
+        + (output_width_stride * output_width_point_value)[:, None]
+    )
+    output_mask = (
+        (in_n_point_value < in_n)[:, None]
+        & (output_c_offset < out_per_group_c)[None, :]
+        & (output_height_point_value < out_height)[:, None]
+        & (output_width_point_value < out_width)[:, None]
+    )
+
+    tl.store(output_pointer, accum, mask=output_mask)
+
+
+class Conv2d(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, input, weight, bias, stride, padding, dilation, groups):
```

Comment:
Dilation is not used at all, so it does not support dilation other than 1?

Reply:
Dilation is not used.

Reply:
I think we should raise an error if dilation > 1.

Reply:
added dilation function